src/lib/comms/sol-coap.c

```diff
@@ -660,13 +664,19 @@ pending_timeout_cb(void *data)
 {
     struct pending_reply *reply = data;
     struct sol_coap_server *server = reply->server;
+    bool callback_result;
 
-    if (reply->cb((void *)reply->data, server, NULL, NULL))
-        return true;
+    SOL_REENTRANT_CALL(reply->reentrant) {
+        callback_result = reply->cb((void *)reply->data, server, NULL, NULL);
+    }
 
-    sol_ptr_vector_remove(&server->pending, reply);
+    if (callback_result && !SOL_REENTRANT_IS_STALE(reply->reentrant)) {
+        return true;
+    }
 
-    pending_reply_free(reply);
+    SOL_REENTRANT_FREE(reply->reentrant) {
+        pending_reply_free(reply);
+    }
```

Comment:
sorry, I've checked the `_FREE()` is bit more than checking the flag, the macro is good!